File: Emergency Messaging Informer/client/src/StompFrame.cpp
#include "../include/StompFrame.h"
#include "StompFrame.h"


using namespace std; // Add this line to use the standard library types without std::

// Default constructor
StompFrame::StompFrame()
    : command(""), headers(), body(""), receiptId(""), isMessage(false) {
}

// Constructor that initializes the frame using a message string
StompFrame::StompFrame(const string &message)
    : command(""), headers(), body(""), receiptId(""), isMessage(false) {
    parseFrame(message);
}

// Parses the STOMP frame from the message string
void StompFrame::parseFrame(const string &message) {
    stringstream ss(message);
    string line;
    string eventDescription;

    // Read lines and split the message
    getline(ss, line);
    command = line; // The first line is the command

    if (command == "MESSAGE") {
        isMessage = true;
    }

    // Parse headers
    while (getline(ss, line) && !line.empty()) {
        size_t colonPos = line.find(":");
        if (colonPos != string::npos) {
            string key = line.substr(0, colonPos);
            string value = line.substr(colonPos + 1);
            if (key == "receipt-id") {
                receiptId = value;
            }
            headers.push_back({key, value});
        }
    }

    // Skip empty line after headers (if present)
    //getline(ss, line);

    // Parse the body until the ^@ delimiter
    while (getline(ss, line) && line != "\0") {
        eventDescription += line + "\n";
    }

    // Remove the last newline from the body (if present)
    if (!eventDescription.empty()) {
        body = eventDescription.substr(0, eventDescription.length() - 1);
    }
}


// Getter methods
string StompFrame::getCommand() const {
    return command;
}

const vector<pair<string, string>>& StompFrame::getHeaders() const {
    return headers;
}

string StompFrame::getBody() const {
    return body;
}

// Setter methods
void StompFrame::setCommand(const string &command) {
    this->command = command;
}

void StompFrame::addHeader(const string &key, const string &value) {
    headers.push_back({key, value});
}

void StompFrame::setBody(const string &body) {
    this->body = body;
}

void StompFrame::addToBody(const string &addToBody) {
    body += addToBody;
}



// Convert the object back to string
string StompFrame::toString() const {
    string ret = command + "\n";
    for (const auto &header : headers) {
        ret += header.first + ":" + header.second + "\n";
    }
    ret += "\n";

    if (!body.empty()) {
        ret += body + "\n";
    }

    ret += "\0";
    return ret;
}

StompFrame StompFrame::createSendFrame(const string &channel, const string &body) 
{
    StompFrame frame;
    frame.setCommand("SEND");
    frame.addHeader("destination",channel);
    frame.setBody(body);

    return frame;
}

StompFrame StompFrame::createConnectFrame(const string& username, const string& password) 
{
    StompFrame frame;
    frame.setCommand("CONNECT");
    frame.addHeader("accept-version","1.2");
    frame.addHeader("host","stomp.cs.bgu.ac.il");
    frame.addHeader("login",username);
    frame.addHeader("passcode",password);
    return frame;
}

StompFrame StompFrame::createDisconnectFrame(const int reciept)
{
    StompFrame frame;
    frame.setCommand("DISCONNECT");
    frame.addHeader("receipt",to_string(reciept));
    return frame;
}
StompFrame StompFrame::createSubscribeFrame(const string &channel, const int subID, const int reciept)
{
    StompFrame frame;
    frame.setCommand("SUBSCRIBE");
    frame.addHeader("destination", channel);
    frame.addHeader("id", to_string(subID));
    frame.addHeader("receipt", to_string(reciept));
    return frame;
}

StompFrame StompFrame::createUnsubscribeFrame(const int subID, const int reciept)
{
    StompFrame frame;
    frame.setCommand("UNSUBSCRIBE");
    frame.addHeader("id", to_string(subID));
    frame.addHeader("receipt", to_string(reciept));
    return frame;
}

File: Emergency Messaging Informer/server/src/main/java/bgu/spl/net/impl/stomp/StompFrame.java
package bgu.spl.net.impl.stomp;

import java.util.LinkedList;
import java.util.List;

//Added file - represents a single frame of whatever type
public class StompFrame {
    private String command;
    private List<String[]> headers;
    private String body;
    private String receiptId; //"" empty string if no receipt
    private boolean isMessage; //If the current frame is a message frame

    public StompFrame(){
        headers= new LinkedList<>();
        receiptId = "";
        isMessage = false;
    }

    public StompFrame(String message){
        headers= new LinkedList<>();
        receiptId = "";
        isMessage = false;
        parseFrame(message);
    }

    public void parseFrame(String message){
        String[] lines = message.toString().split("\n"); // Split the message by newlines
        command = lines[0]; // The first line is the command
        if(command.compareTo("MESSAGE")==0){
            isMessage=true;
        }

        StringBuilder bodyBuilder = new StringBuilder();

        int i = 1;
        // Parse headers
        while (i < lines.length && !lines[i].isEmpty()) { // Empty line indicates end of headers
            String line = lines[i];
            if (line.contains(":")) {
                String[] headerParts = line.split(":", 2); // Split into key and value
                if(headerParts[0].compareTo("receipt-id")==0){
                    receiptId = headerParts[1];
                }
                headers.add(new String[]{headerParts[0].trim(), headerParts[1].trim()});
            }
            i++;
        }
        // Skip empty line (if present) after headers
        if (i < lines.length && lines[i].isEmpty()) {
            i++;
        }

        // Parse body until the ^@ delimiter
        while (i < lines.length && !lines[i].equals("^@")) {
            bodyBuilder.append(lines[i]).append("\n");
            i++;
        }

        // Remove the last newline from the body (if present)
        body = bodyBuilder.length() > 0
                ? bodyBuilder.substring(0, bodyBuilder.length() - 1)
                : null;
    }

    public String getCommand(){
        return command;
    }

    public List<String[]> getHeaders(){
        return headers;
    }

    public String getBody(){
        return body;
    }

    public void setCommand(String command){
        this.command=command;
    }
    public void addHeader(String[] header){
        headers.add(header);
    }
    public void setBody(String body){
        this.body=body;
    }

    public void addToBody(String addToBody){
        body = body+addToBody;
    }

    public String toString(){
        String ret="";
        ret = ret+command+"\n";
        for(String[] header : headers){
            ret = ret + header[0] + ":" + header[1] + "\n";
        }
        ret = ret + "\n";
        if(body!=null && body.compareTo("")!=0){
            ret = ret + body + "\n";
        }
        ret = ret+"^@";
        return ret;
    }

    public static StompFrame createRecieptFrame(int receiptId){
        StompFrame frame = new StompFrame("RECEIPT\nreceipt-id:"+receiptId+"\n\n");
        return frame;
    }

    public static StompFrame createMessageFrame(int messageId, String channel, String frameBody){
        StompFrame frame = new StompFrame();
        frame.setCommand("MESSAGE");
        frame.addHeader(new String[] {"message-id",String.valueOf(messageId)});
        frame.addHeader(new String[] {"destination",channel});
        frame.setBody(frameBody);
        return frame;
    }


    public static StompFrame createMessageFrame(int messageId, String channel, int subscriptionId, String frameBody){
        StompFrame frame = new StompFrame("MESSAGE\nsubscription:"+subscriptionId+"\nmessage-id:"
                                            +messageId+"\ndestination:/"+channel+"\n\n"+frameBody+"\n");
        return frame;
    }

    public static StompFrame createErrorFrame(String shortDescription, String detailedDescription, int receiptId, String malformedFrame){
        String reciept = "";
        if (receiptId!=-1) {
            reciept = "reciept-id:"+receiptId+"\n";
        }
        StompFrame frame = new StompFrame("ERROR\n"+reciept+"message:malformed frame recieved\n\nThe message:"
                                            +shortDescription+"\n----\n"+malformedFrame+"\n----\n"
                                            +detailedDescription+"\n");
        return frame;
    }

    public static StompFrame createConnectedFrame(){
        StompFrame frame = new StompFrame("CONNECTED\nversion:1.2\n\n");
        return frame;
    }

    public String getReceiptId(){
        return receiptId;
    }

    public boolean isMessageFrame(){
        return isMessage;
    }
}
